Make the `llnl` package independent of `spack`

alalazo opened the pull request:
closes #34767

This PR completely separates the `llnl` Python package from `spack`. The new package `llnl.syscmd` is a good candidate for extraction from Spack, similarly to `sbang` and `archspec`.

Modifications:
- [x] Moved `spack.util.environment` and `spack.util.executable` into `llnl.syscmd`

The current dependency graph is:

![llnl](https://github.com/spack/spack/assets/4199709/b74e05be-4263-4969-8287-e801cc3cb744)

alalazo committed:
Remove dependency of spack.util.executable from spack.error

alalazo committed:
Remove local import from spack.util.executable

haampie commented:
imho this is more breaking than necessary...

vincentmr commented on var/spack/repos/builtin/packages/py-pennylane-lightning/package.py:
```diff
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
-
+import spack.build_systems.cmake
```
Here `spack.build_systems.cmake` is imported, but we are still using `spack.build_systems.cmake.CMakeBuilder` at [line 62](bbc7c0a9fcb60b7f6a12fb5e0636d625ec1c3e60b212cc65ed952492abe9eea3R62). Was the goal to use a more targeted import as in `var/spack/repos/builtin/packages/py-pennylane-lightning-kokkos/package.py`? Should we make that change in other files like GROMACS'?

ChristianTackeGSI commented on var/spack/repos/builtin/packages/git/package.py:
```diff
@@ -6,8 +6,10 @@
 
+from llnl.syscmd import is_system_path
+
+import spack.fetch_strategy
```
(Not a maintainer of git, just a random note)

Why is `spack.fetch_strategy` needed now? Why wasn't it needed before?

If it was needed before anyhow, maybe put that in another PR?

alalazo replied:
The import system in `python` is flawed (personal opinion), and you might get away with not importing required packages, if they are transitively imported by something else. This is similar to transitive:
```
#include "xxxxx"
```
that you get from header files in C or C++. Since I polished the circular imports of a Python package in this PR, this revealed a few places where "xxxxx" was needed and not explicitly imported.

ChristianTackeGSI replied:
I *personally* would vote for putting this sort of "fixes" into its own "cleanup / fixup / improve" PR. Something like "We learned, that more imports are really needed".

As I haven't done any spack tool development in a long time, please consider this only as something to possibly "ponder".

ChristianTackeGSI commented on var/spack/repos/builtin/packages/gnupg/package.py:
```diff
@@ -3,8 +3,9 @@
+from llnl.syscmd import is_system_path
+
 from spack.package import *
-from spack.util.environment import is_system_path
```
Seeing that `is_system_path` is changed in many `package.py` files, I wonder two things:

* Why isn't this function part of some `llnl*path` library instead?
* If it's moved there, this could be factored into a smaller PR also?

ChristianTackeGSI commented on var/spack/repos/builtin/packages/mesa18/package.py:
```diff
@@ -44,9 +44,7 @@ class Mesa18(AutotoolsPackage):
-        values=spack.variant.DisjointSetsOfValues(
-            ("none",), ("auto",), ("avx", "avx2", "knl", "skx")
-        )
+        values=DisjointSetsOfValues(("none",), ("auto",), ("avx", "avx2", "knl", "skx"))
```
This change looks quite unrelated to me?

alalazo replied:
It is related. We didn't import `spack.variant` (see https://github.com/spack/spack/pull/40262#discussion_r1380648533), but we are forwarding the name `DisjointSetOfValues` to `spack.package` which is imported at the top. `spack.package` is a collection of directives and other methods, classes etc. that are frequently used in package recipes.

So I fixed the missing import, by referring to the same name imported from elsewhere.

ChristianTackeGSI replied:
Okay, got it. The same as above: Maybe put this in a preparatory PR that gets merged before this one?

ChristianTackeGSI reviewed (commented):
I was pinged to review this. I don't think, that I am the right person for those "system changes". (I was pinged because of changes in some package.py I assume).

That said, I have a few random comments.

haampie commented:
Not immediately a fan of the new imports in `package.py` files. It's roughly this list:

```python
from llnl.syscmd import which, which_string, is_system_path, from_sourcing_file, Executable, EnvironmentModifications, ProcessError, set_env
import llnl.syscmd
import spack.compiler
import spack.error
import spack.platforms
import spack.version
```

Better to carefully design the `spack.package` API instead of sprinkling around explicit imports in package.py files, which will only break forward compatibility of the builtin repo.

All in all this PR has a net negative impact on the 0.21 release, since it requires folks to update their internal repo, and their internal repo won't be usable with both old and new Spack versions...

I would suggest not to merge this and do more careful refactoring.

alalazo commented:
> Better to carefully design the spack.package API instead of sprinkling around explicit imports in package.py files, which will only break forward compatibility of the builtin repo.

That doesn't solve the issue that we have those imports right now in builtin. Refactoring what we export in `spack.package` is not difficult, but if we want to remove at some point the circular deps, how should we treat the imports that starts erroring?

haampie commented:
A few common names are missing in `spack.package.*`, and this refactor imports them explicitly. Things worked "by accident"

My suggestion is to 

1. Add those names to `spack.package.*`
2. Drop any fully qualified name `(spack|llnl).popular_module.popular_name` in all `package.py`s and replace with `popular_name` re-exported from `spack.package`
3. Leave that until Spack 0.22, while "deprecating" direct use internal API (how? dunno yet)
4. Only move modules in Spack 0.23, cause then people can at least update their repo and have it work with current and previous Spack (0.23 & 0.22), instead of having to write ugly `try import ... except ImportError` etc.

alalazo committed:
Extract a couple of functions from spack.util.environment_modifications

Extracting these functions makes the module independent of
spack.util.executable

alalazo committed:
Package modifications

alalazo committed:
Move environment modifications to llnl.syscmd

alalazo committed:
Move executable to llnl.syscmd

alalazo committed:
Move sourcing to llnl.syscmd

alalazo committed:
Fix a few missed imports

alalazo committed:
Forward attribute lookup from deleted modules

Any attribute lookup from spack.util.executable or
spack.util.environment is forwarded to llnl.syscmd.

A FutureWarning is emitted on import.

alalazo committed:
spack.package: add back transitive imports for backward compatibility

alalazo commented:
Latest commit should be backward compatible wrt to packages

alalazo committed:
style: use black v24.X
